### ISSUE TITLE
"Assertion Failed" crash in peep.c

### ISSUE THREAD

**Issue by evanjhopkins** [OpenRCT2/OpenRCT2#2417]
Sorry if this is a duplicate. My searches didn't turn up any other issues about this.

Platform: Windows
Version: 93e0bc5
Context: Happens online in game seemingly at random
Steps to reproduce: Unknown

![image](https://cloud.githubusercontent.com/assets/590209/11516517/86577d7e-9838-11e5-9efe-785b864fc5d5.png)

**Comment by duncanspumpkin** [comment]
Any information on what you were doing to cause this?

**Comment by evanjhopkins** [comment]
Happens seemingly at random during online play. I'm trying to figure out how to reproduce the error on purpose but no luck so far.

**Comment by Gymnasiast** [comment]
I think it might the assertion here: https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/peep/peep.c#L6592

(Offtopic and just out of interest: do you speak Dutch? Your screenshot has Dutch buttons but your profile and website seem to suggest you're American.)

**Comment by evanjhopkins** [comment]
I do not speak dutch unfortunately. The error was posted on a facebook group and a second user reported seeing the same error.

**Comment by zsilencer** [comment]
I got this once, it happens because guest_surface_path_finding sometimes calls peep_move_one_tile with randDirection as -1 or 255.

**Comment by janisozaur** [comment]
@zsilencer are you sure? `guest_surface_path_finding` looks fairly safe to me. I'd rather suspect `guest_path_finding`, line 7258:
```
	direction = bitscanforward(edges);
	// IF only one edge to choose from
	if ((edges & ~(1 << direction)) == 0) {
		return peep_move_one_tile(direction, peep);
	}
```

**Comment by janisozaur** [comment]
I don't think it should have `multiplayer` label, I ran into it once as well, long time ago, and I wasn't playing multi.

**Comment by HaasJona** [comment]
I agree with @zsilencer. guest_surface_path_finding seems to be suspicious.

Assume `randDirection` equals `1` at the start, then `randDirection++` equals `2` which is unchanged on `randDirection &= 3;`. Assuming there are always fences in the way, the following code gets executed:

	randDirection -= 2;
	randDirection &= 3;

which results in randDirection being 0.

And then we call `randDirection--;` at the end which results in an underflow if rand_backwards is false.

But guest_path_finding might also be problematic.

**Comment by janisozaur** [comment]
You might be right. In this case
```
randDirection &= 3;
```
Just before a call to `peep_move_one_tile` in line 6673 should fix it.

**Comment by IntelOrca** [comment]
I think both cases are suspicious.
```guest_surface_path_finding```: subtracting 1 and now ANDing it by 3 again
```guest_path_finding```: if edges = 0, -1 is returned from bitscanforward

**Comment by duncanspumpkin** [comment]
https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/peep/peep.c#L7258 can't be -1 as the code is written so that there is always an edge.

**Comment by duncanspumpkin** [comment]
This should be in the hotfix as well `pre-release-0.0.3.1`

**Comment by AlexanderLaheij** [comment]
Never had this isseu before. Since new update happens a lot. I think it has to do with guests.

**Comment by IntelOrca** [comment]
@Overv it might have happened before, we only added the assertion recently (otherwise it could of just been a hidden bug)

**Comment by IntelOrca** [comment]
@duncanspumpkin What if there is only one edge and its the direction that is removed from edges:
https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/peep/peep.c#L7255

**Comment by duncanspumpkin** [comment]
Then this wouldn't occur https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/peep/peep.c#L7250

**Comment by IntelOrca** [comment]
Hmm, I do not like ```!``` being used in that context - just like strcmp.

**Comment by duncanspumpkin** [comment]
It was written like that so that it matched the original as close as possible. The idea is remove direction that is behind you only if that won't cause you to have no directions to turn to. var_78 should really be labelled as last turn direction but I remember there being reasons as to why I didn't.

**Comment by IntelOrca** [comment]
What I find confusing is that it isn't checking for a single direction, its checking for all but one direction and then negating that.

**Comment by duncanspumpkin** [comment]
Fixed in the pre-release-0.0.3.1

**Comment by duncanspumpkin** [comment]
I'm reopening this so that people don't re submit. Will be closed proper when pre-release merged.
